File: day-2/rust/src/solution/wrapping_paper.rs
use crate::solution::model::dimension::Dimension;

pub fn part_1(contents: &Vec<String>) -> Result<usize, String> {
  let dimensions = parse_dimensions(contents);
  if dimensions.len() == 0 {
    return Ok(0);
  }
  let mut total_surface_area = 0;
  for dim in dimensions {
    total_surface_area += dim.total_wrapping_paper()
  }
  Ok(total_surface_area)
}

pub fn part_2(contents: &Vec<String>) -> Result<usize, String> {
  let dimensions = parse_dimensions(contents);
  if dimensions.len() == 0 {
    return Ok(0);
  }
  let mut total_ribbon = 0;
  for dim in dimensions {
    total_ribbon += dim.total_ribbon();
  }
  Ok(total_ribbon)
}

fn parse_dimensions(contents: &Vec<String>) -> Vec<Dimension> {
  let mut dimensions = Vec::new();
  for dim in contents {
    dimensions.push(Dimension::from_string(&dim))
  }
  dimensions
}

File: day-2/go/solution/wrapping_paper.go
package wrapping_paper

import (
	"fmt"

	dimension "github.com/justint-rl/advent-of-code-2015/day-2/go/model/dimension"
)

type WrappingPaper interface {
	Part1(data []string) int64
	Part2(data []string) int64
}

type wrappingPaperImpl struct{}

func New() WrappingPaper {
	return &wrappingPaperImpl{}
}

func (w *wrappingPaperImpl) Part1(data []string) int64 {
	totalWrappingPaper := int64(0)
	for _, d := range data {
		dim, err := dimension.FromString(d)
		if err != nil {
			fmt.Println("Error converting string to Dimension:", err)
		}
		totalWrappingPaper += (dim.SurfaceArea() + dim.SmallestSideArea())
	}
	return totalWrappingPaper
}

func (w *wrappingPaperImpl) Part2(data []string) int64 {
	return 0
}
